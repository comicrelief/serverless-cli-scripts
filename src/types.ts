/**
 * CLI Context
 */
export interface CLIContext {
  stage?: string;
}
