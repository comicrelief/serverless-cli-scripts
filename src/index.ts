export { runner } from './runner';
export { deploy } from './commands/deploy';
