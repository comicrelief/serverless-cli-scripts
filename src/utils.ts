import { spawn as childProcessSpawn } from 'child_process';

import consola from 'consola';

/**
 * Spawns a subprocess and observes it
 *
 * @param command
 * @param commandArgs
 */
export const spawn = async (command: string, commandArgs?: string[], quiet = false): Promise<number> => {
  return new Promise((resolve, reject) => {
    const childProcess = childProcessSpawn(command, commandArgs);

    if (!quiet) {
      consola.info(`Executing: '${command} ${(commandArgs || []).join(' ')}'`);

      childProcess.stdout.on('data', (data) => {
        consola.info(data.toString());
      });

      childProcess.stderr.on('data', (data) => {
        consola.error(data.toString());
      });
    }

    childProcess.on('exit', (code) => {
      const exitCode = code || 0;

      if (exitCode !== 0) {
        return reject(exitCode);
      }

      resolve(exitCode);
    });

    // Makes sure that if we CTRL+C the script
    // we also kill the spawned child process
    process.on('SIGINT', () => {
      childProcess.kill();
      process.exit();
    });
  })
};
