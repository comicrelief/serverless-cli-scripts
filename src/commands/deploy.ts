import prompts from 'prompts';

import type { CLIContext } from '../types';
import * as utils from '../utils';

export const CONFIRM_DEPLOY_TO_PROD = 'deploy-to-production';

/**
 * Asks the user to confirm
 * a deploy to production
 */
export const confirmDeployToProduction = async (): Promise<string> => {
  const { confirmProd } = await prompts({
    type: 'text',
    name: 'confirmProd',
    message: `Please type '${CONFIRM_DEPLOY_TO_PROD}' to confirm deploy to production.`,
  });

  return confirmProd;
};

/**
 * Deploys to the given stage
 *
 * @param context
 */
export const deploy = async (context: CLIContext): Promise<number> => {
  const stage = context.stage || 'dev';

  if (stage === 'production') {
    const confirm = await confirmDeployToProduction();

    if (confirm !== CONFIRM_DEPLOY_TO_PROD) {
      throw new Error('Bailing out.');
    }
  }

  const command = 'yarn';
  const commandArgs = ['sls', 'deploy', '--stage', stage];

  return utils.spawn(command, commandArgs);
};
