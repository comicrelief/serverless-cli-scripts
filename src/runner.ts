/* eslint-disable import/no-extraneous-dependencies */
import consola from 'consola';

import { deploy } from './commands/deploy';

export const commands = {
  deploy,
};

export interface CLIArgs {
  [x: string]: unknown;
  _: (string | number)[];
  $0: string;
  stage?: string;
}

export interface Commands {
  [name: string]: (context: CLIArgs) => Promise<void>
}

export interface CLIRunnerOptions<T extends Commands> {
  getArgs: () => CLIArgs,
  loadEnv: (stage: string) => Promise<void>;
  commands: T
}

/**
 * Runs as a main function
 *
 * @param options
 */
export const runner = async <T extends Commands>(options: CLIRunnerOptions<T>): Promise<void> => {
  const context = options.getArgs();

  const [command] = context._;
  const executor = options.commands[command as keyof T];

  if (!executor) {
    throw new Error(`'${command}' is not a valid command.`);
  }

  consola.info(`Running: ${command}`);

  options.loadEnv(context.stage || '');

  await executor(context);
};
