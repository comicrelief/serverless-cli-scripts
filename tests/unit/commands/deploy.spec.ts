import * as deploy from '../../../src/commands/deploy';
import * as utils from '../../../src/utils';

describe('unit.commands.deploy', () => {
  afterEach(() => jest.resetAllMocks());

  describe('deploy', () => {
    describe('with stage: production', () => {
      it('fails with any text', () => {
        jest.spyOn(deploy, 'confirmDeployToProduction').mockImplementation(() => Promise.resolve('abc'));
        jest.spyOn(utils, 'spawn').mockImplementation(() => Promise.resolve(0));

        const promise = deploy.deploy({ stage: 'production' });

        expect(promise).rejects.toThrowErrorMatchingSnapshot();
        expect(deploy.confirmDeployToProduction).toHaveBeenCalledTimes(1);
      });

      it(`accepts '${deploy.CONFIRM_DEPLOY_TO_PROD}'`, () => {
        jest.spyOn(deploy, 'confirmDeployToProduction').mockImplementation(() => Promise.resolve(deploy.CONFIRM_DEPLOY_TO_PROD));
        jest.spyOn(utils, 'spawn').mockImplementation(() => Promise.resolve(0));

        const promise = deploy.deploy({ stage: 'production' });

        expect(promise).resolves.toMatchSnapshot();
        expect(deploy.confirmDeployToProduction).toHaveBeenCalledTimes(1);
      });
    });

    ['dev', 'staging'].forEach((stage) => {
      describe(`with stage: ${stage}`, () => {
        it('throws with an exit code > 0', () => {
          jest.spyOn(deploy, 'confirmDeployToProduction').mockImplementation(() => Promise.resolve(deploy.CONFIRM_DEPLOY_TO_PROD));
          // eslint-disable-next-line prefer-promise-reject-errors
          jest.spyOn(utils, 'spawn').mockImplementation(() => Promise.reject(1));

          const promise = deploy.deploy({ stage });

          expect(promise).rejects.toThrowErrorMatchingSnapshot();
          expect(deploy.confirmDeployToProduction).toHaveBeenCalledTimes(0);
        });

        it('returns status code 0', () => {
          jest.spyOn(deploy, 'confirmDeployToProduction').mockImplementation(() => Promise.resolve(deploy.CONFIRM_DEPLOY_TO_PROD));
          // eslint-disable-next-line prefer-promise-reject-errors
          jest.spyOn(utils, 'spawn').mockImplementation(() => Promise.resolve(0));

          const promise = deploy.deploy({ stage });

          expect(promise).resolves.toMatchSnapshot();
          expect(deploy.confirmDeployToProduction).toHaveBeenCalledTimes(0);
        });
      });
    });
  });
});
