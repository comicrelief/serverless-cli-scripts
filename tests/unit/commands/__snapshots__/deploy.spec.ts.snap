// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit.commands.deploy deploy with stage: dev returns status code 0 1`] = `0`;

exports[`unit.commands.deploy deploy with stage: dev throws with an exit code > 0 1`] = `undefined`;

exports[`unit.commands.deploy deploy with stage: production accepts 'deploy-to-production' 1`] = `0`;

exports[`unit.commands.deploy deploy with stage: production fails with any text 1`] = `"Bailing out."`;

exports[`unit.commands.deploy deploy with stage: staging returns status code 0 1`] = `0`;

exports[`unit.commands.deploy deploy with stage: staging throws with an exit code > 0 1`] = `undefined`;
